src/application-manager.coffee

```diff
@@ -0,0 +1,995 @@
+			_.some(available, (availableImage) => @images.isSameImage(availableImage, targetImage)) and
+				!_.some(availableWithoutIds, (img) -> _.isEqual(img, targetImage))
+
+		deltaSources = _.map imagesToDownload, (image) =>
+			return @bestDeltaSource(image, available)
+		proxyvisorImages = @proxyvisor.imagesInUse(current, target)
+		imagesToRemove = _.filter availableAndUnused, (image) =>
+			notUsedForDelta = !_.includes(deltaSources, image.name)
+			notUsedByProxyvisor = !_.some proxyvisorImages, (proxyvisorImage) => @images.isSameImage(image, { name: proxyvisorImage })
+			return notUsedForDelta and notUsedByProxyvisor
+		return { imagesToSave, imagesToRemove }
+
+	_inferNextSteps: (cleanupNeeded, availableImages, downloading, supervisorNetworkReady, current, target, ignoreImages, { localMode, delta }) =>
+		Promise.try =>
+			delta = checkTruthy(delta)
+			if checkTruthy(localMode)
+				target = _.cloneDeep(target)
+				target.local.apps = _.mapValues target.local.apps ? {}, (app) ->
+					app.services = []
+					return app
+				ignoreImages = true
+			currentByAppId = current.local.apps ? {}
+			targetByAppId = target.local.apps ? {}
+			nextSteps = []
+			if !supervisorNetworkReady
+				nextSteps.push({ action: 'ensureSupervisorNetwork' })
+			else
+				if !ignoreImages and _.isEmpty(downloading)
+					if cleanupNeeded
+						nextSteps.push({ action: 'cleanup' })
+					{ imagesToRemove, imagesToSave } = @_compareImages(current, target, availableImages)
+					for image in imagesToSave
+						nextSteps.push({ action: 'saveImage', image })
+					if _.isEmpty(imagesToSave)
+						for image in imagesToRemove
+							nextSteps.push({ action: 'removeImage', image })
+				# If we have to remove any images, we do that before anything else
+				if _.isEmpty(nextSteps)
+					allAppIds = _.union(_.keys(currentByAppId), _.keys(targetByAppId))
+					for appId in allAppIds
+						nextSteps = nextSteps.concat(@_nextStepsForAppUpdate(currentByAppId[appId], targetByAppId[appId], availableImages, downloading))
+			newDownloads = _.filter(nextSteps, (s) -> s.action == 'fetch').length
+			if !ignoreImages and delta and newDownloads > 0
+				downloadsToBlock = downloading.length + newDownloads - constants.maxDeltaDownloads
+				while downloadsToBlock > 0
+					_.pull(nextSteps, _.find(nextSteps, action: 'fetch'))
+					downloadsToBlock -= 1
+			if !ignoreImages and _.isEmpty(nextSteps) and !_.isEmpty(downloading)
+				nextSteps.push({ action: 'noop' })
+			return _.uniqWith(nextSteps, _.isEqual)
+
+	stopAll: ({ force = false, skipLock = false } = {}) =>
+		@services.getAll()
+		.map (service) =>
+			@_lockingIfNecessary service.appId, { force, skipLock }, =>
+				@services.kill(service, { removeContainer: false, wait: true })
+				.then =>
+					delete @_containerStarted[service.containerId]
+
+	_lockingIfNecessary: (appId, { force = false, skipLock = false } = {}, fn) =>
+		if skipLock
+			return Promise.try(fn)
```

Comment:
This could use `Promise.method` and then `return fn()`, but it's a take it or leave it imo